sharding/indexHashedNodesCoordinator.go

```diff
@@ -239,37 +351,140 @@ func (ihgs *indexHashedNodesCoordinator) GetSelectedPublicKeys(selection []byte,
-func (ihgs *indexHashedNodesCoordinator) GetValidatorsIndexes(publicKeys []string) []uint64 {
-	validatorsPubKeys := ihgs.GetAllValidatorsPublicKeys()
+func (ihgs *indexHashedNodesCoordinator) GetValidatorsIndexes(
+	publicKeys []string,
+	epoch uint32,
+) ([]uint64, error) {
 	signersIndexes := make([]uint64, 0)
 
+	validatorsPubKeys, err := ihgs.GetAllValidatorsPublicKeys(epoch)
+	if err != nil {
+		return nil, err
+	}
+
+	ihgs.mutNodesConfig.RLock()
+	nodesConfig := ihgs.nodesConfig[epoch]
+	ihgs.mutNodesConfig.RUnlock()
+
 	for _, pubKey := range publicKeys {
-		for index, value := range validatorsPubKeys[ihgs.shardId] {
+		for index, value := range validatorsPubKeys[nodesConfig.shardId] {
 			if bytes.Equal([]byte(pubKey), value) {
 				signersIndexes = append(signersIndexes, uint64(index))
 			}
 		}
 	}
 
-	return signersIndexes
+	if len(publicKeys) != len(signersIndexes) {
+		log.Error("len pubKeys = %d, len signersIndexes = %d", len(publicKeys), len(signersIndexes))
+		return nil, ErrNotInvalidNumberPubKeys
+	}
+
+	return signersIndexes, nil
+}
+
+// EpochStartPrepare wis called when an epoch start event is observed, but not yet confirmed/committed.
+// Some components may need to do some initialisation on this event
+func (ihgs *indexHashedNodesCoordinator) EpochStartPrepare(metaHeader data.HeaderHandler) {
+	randomness := metaHeader.GetRandSeed()
+	newEpoch := metaHeader.GetEpoch()
+
+	if newEpoch != ihgs.currentEpoch+1 {
+		return
+	}
+
+	ihgs.mutNodesConfig.RLock()
+	nodesConfig, ok := ihgs.nodesConfig[ihgs.currentEpoch]
+	ihgs.mutNodesConfig.RUnlock()
+
+	if !ok {
+		log.Error("no configured epoch found")
+		return
+	}
+
+	// TODO: update the new nodes and leaving nodes as well
+	shufflerArgs := ArgsUpdateNodes{
+		Eligible: nodesConfig.eligibleMap,
+		Waiting:  nodesConfig.waitingMap,
+		NewNodes: make([]Validator, 0),
+		Leaving:  make([]Validator, 0),
+		Rand:     randomness,
+		NbShards: nodesConfig.nbShards,
+	}
+
+	eligibleMap, waitingMap, _ := ihgs.shuffler.UpdateNodeLists(shufflerArgs)
+
+	_ = ihgs.SetNodesPerShards(eligibleMap, waitingMap, newEpoch)
+}
+
+// EpochStartAction is called upon a start of epoch event.
+// NodeCoordinator has to get the nodes assignment to shards using the shuffler.
+func (ihgs *indexHashedNodesCoordinator) EpochStartAction(hdr data.HeaderHandler) {
+	newEpoch := hdr.GetEpoch()
+	epochToRemove := int32(newEpoch) - nodeCoordinatorStoredEpochs
+	needToRemove := epochToRemove >= 0
+	ihgs.currentEpoch = newEpoch
+
+	ihgs.mutNodesConfig.Lock()
+	if needToRemove {
+		for epoch := range ihgs.nodesConfig {
+			if epoch <= uint32(epochToRemove) {
+				delete(ihgs.nodesConfig, epoch)
+			}
+		}
+	}
+	ihgs.mutNodesConfig.Unlock()
 }
 
-func (ihgs *indexHashedNodesCoordinator) expandEligibleList(shardId uint32) []Validator {
+func (ihgs *indexHashedNodesCoordinator) expandEligibleList(validators []Validator, mut *sync.RWMutex) []Validator {
 	//TODO implement an expand eligible list variant
```

Comment:
The expansion of the eligible list is only related to the effects rating has on the chances of being selected. 

Since the computation of consensus group is handled by the nodes coordinator, then the logic should be here as well. 

The expansion will finally be done only once per checkpoint (when rating modifications take effect) and the cached one will be used for consensus group computation. 
There are still some changes pending to complete this, but not part of this PR.